File: swift/DesignLinkedList.swift
// https://leetcode.com/problems/design-linked-list/submissions/

class MyLinkedList {
    
    class Node {
        var val: Int?
        var next: Node?
        
        init(val: Int? = nil, next: Node? = nil) {
            self.val = val
            self.next = next
        }
    }
    
    private var head: Node

    init() {
        head = Node()
    }
    
    func get(_ index: Int) -> Int {
        var target = head.next
        var counter = 0
        while counter < index {
            target = target?.next
            counter += 1
        }
        return target?.val ?? -1
    }
    
    func addAtHead(_ val: Int) {
        head.next = Node(val: val, next: head.next)
    }
    
    func addAtTail(_ val: Int) {
        var tail = head
        while tail.next != nil {
            tail = tail.next!
        }
        tail.next = Node(val: val)
    }
    
    func addAtIndex(_ index: Int, _ val: Int) {
        guard index >= 0 else { return }
        
        var target: Node? = head
        var counter = 0
        while counter < index {
            target = target?.next
            counter += 1
        }
        target?.next = Node(val: val, next: target?.next)
    }
    
    func deleteAtIndex(_ index: Int) {
        guard index >= 0 else { return }
        
        var prev: Node? = head
        var counter = 0
        while counter < index {
            prev = prev?.next
            counter += 1
        }
        
        prev?.next = prev?.next?.next
    }
}

File: c/DesignLinkedList.c
// https://leetcode.com/problems/design-linked-list/

typedef struct List{
    int val;
    struct List * next;
} MyLinkedList;

MyLinkedList* myLinkedListCreate() {
    MyLinkedList *list = malloc(sizeof(MyLinkedList));
    list->val = -1;
    list->next = NULL;
    return list;
}

int myLinkedListGet(MyLinkedList* obj, int index) {
    MyLinkedList* tmp = obj;
    int i = 0;
    while (i++ <= index && tmp) {
        tmp = tmp->next;
    }
    return tmp ? tmp->val : -1;
}

void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    MyLinkedList *list = malloc(sizeof(MyLinkedList));
    list->val = val;
    list->next = obj->next;
    obj->next = list;
}

void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    MyLinkedList *tail = obj;
    while (tail->next != NULL)
        tail = tail->next;
    MyLinkedList *list = malloc(sizeof(MyLinkedList));
    list->val = val;
    list->next = NULL;
    tail->next = list;
}

void myLinkedListAddAtIndex(MyLinkedList* obj, int index, int val) {
    MyLinkedList* tmp = obj;
    int i = 0;
    while (i++ < index && tmp)
        tmp = tmp->next;
        
    if (tmp) {
        MyLinkedList *list = malloc(sizeof(MyLinkedList));
        list->val = val;
        list->next = tmp->next;
        tmp->next = list;
    }
}

void myLinkedListDeleteAtIndex(MyLinkedList* obj, int index) {
    MyLinkedList *tmp = obj, *loc;;
    int i = 0;
    while(i++ < index && tmp)
        tmp = tmp->next;
    
    if (tmp)  
        if(tmp->next && tmp->next->next != NULL) {
            loc =tmp->next;
            tmp->next = tmp->next->next;
            free(loc);
        } else if(tmp->next && tmp->next->next == NULL) {
            loc = tmp->next;
            free(loc);
            tmp->next = NULL;
        }
}

void myLinkedListFree(MyLinkedList* obj) {
    MyLinkedList* tmp = obj;
    
    while (tmp->next) {
        MyLinkedList* loc = tmp->next;
        tmp->next = tmp->next->next;
        free(loc);
    }
}

